pkg/loop/ccip_commit.go

```diff
@@ -0,0 +1,92 @@
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/loop/internal/goplugin"
+	"github.com/smartcontractkit/chainlink-common/pkg/loop/internal/reportingplugin/ccip"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+)
+
+// CCIPCommitLOOPName is the name for [types.CCIPCommitFactoryGenerator]/[NewCommitLOOP].
+const CCIPCommitLOOPName = "ccip_commit"
+
+func PluginCCIPCommitHandshakeConfig() plugin.HandshakeConfig {
+	return plugin.HandshakeConfig{
+		MagicCookieKey:   "CL_PLUGIN_CCIP_COMMIT_MAGIC_COOKIE",
+		MagicCookieValue: "5a2d1527-6c0f-4c7e-8c96-00aa4bececd2",
+	}
+}
+
+type CommitLoop struct {
+	plugin.NetRPCUnsupportedPlugin
+
+	BrokerConfig
+
+	PluginServer types.CCIPCommitFactoryGenerator
+
+	pluginClient *ccip.CommitLOOPClient
+}
+
+func (p *CommitLoop) GRPCServer(broker *plugin.GRPCBroker, server *grpc.Server) error {
+	return ccip.RegisterCommitLOOPServer(server, broker, p.BrokerConfig, p.PluginServer)
+}
+
+// GRPCClient implements [plugin.GRPCPlugin] and returns the pluginClient [types.CCIPCommitFactoryGenerator], updated with the new broker and conn.
+func (p *CommitLoop) GRPCClient(_ context.Context, broker *plugin.GRPCBroker, conn *grpc.ClientConn) (interface{}, error) {
+	if p.pluginClient == nil {
+		p.pluginClient = ccip.NewCommitLOOPClient(broker, p.BrokerConfig, conn)
+	} else {
+		p.pluginClient.Refresh(broker, conn)
+	}
+
+	return types.CCIPCommitFactoryGenerator(p.pluginClient), nil
+}
+
+func (p *CommitLoop) ClientConfig() *plugin.ClientConfig {
+	clientConfig := &plugin.ClientConfig{
+		HandshakeConfig: PluginCCIPCommitHandshakeConfig(),
+		Plugins:         map[string]plugin.Plugin{CCIPCommitLOOPName: p},
+	}
+	return ManagedGRPCClientConfig(clientConfig, p.BrokerConfig)
+}
+
+var _ ocrtypes.ReportingPluginFactory = (*CommitFactoryService)(nil)
+
+// CommitFactoryService is a [types.Service] that maintains an internal [types.CCIPCommitFactoryGenerator].
+type CommitFactoryService struct {
+	goplugin.PluginService[*CommitLoop, types.ReportingPluginFactory]
+}
+
+// NewCommitService returns a new [*CommitFactoryService].
+// cmd must return a new exec.Cmd each time it is called.
+func NewCommitService(lggr logger.Logger, grpcOpts GRPCOpts, cmd func() *exec.Cmd, provider types.CCIPCommitProvider) *CommitFactoryService {
+	newService := func(ctx context.Context, instance any) (types.ReportingPluginFactory, error) {
+		plug, ok := instance.(types.CCIPCommitFactoryGenerator)
+		if !ok {
+			return nil, fmt.Errorf("expected CCIPCommitFactoryGenerator but got %T", instance)
+		}
+		return plug.NewCommitFactory(ctx, provider)
+	}
+	stopCh := make(chan struct{})
+	lggr = logger.Named(lggr, "CCIPCommitService")
+	var efs CommitFactoryService
+	broker := BrokerConfig{StopCh: stopCh, Logger: lggr, GRPCOpts: grpcOpts}
+	efs.Init(CCIPCommitLOOPName, &CommitLoop{BrokerConfig: broker}, newService, lggr, cmd, stopCh)
```

Comment:
this odd that you can Init by value rather than a pointer. i would have guessed Init is mutating the struct. can you double check want Init does?

cfs := &CommitFactoryService{}
...
cfs.Init(..)
...
return cfs

as an alternative

Reply:
It is mutating the struct, good call. I think it's fine because it's a pointer receiver. I added 

```
	t.Run("service not nil", func(t *testing.T) {
		require.NotPanics(t, func() { commit.Name() })
	})
```

to the unit test suite to demonstrate.